### ISSUE TITLE
Avoid adding a new windows firewall exemption every time an application is updated

### ISSUE THREAD

**Issue by peters** [Squirrel/Squirrel.Windows#1052]
The problem is that every time an application is updated, the user has to add (accept) a new firewall exemption. This is probably fine for most users, but this application is running in a unattended mode (Corporate 💵 ).

I have two ideas on how we can get around this "limitation". Upon first install an exemption will have to be accepted by the user (Windows default prompt anyways):

1 - Host the CLR inside a stub executable and load the target application into memory. This would be a opt-in feature during releasify and only work for managed applications. 

2 - Use .exe.config trickery rewriting the DLL load path (only works for managed dlls), modify environment "PATH" variable to include "app-X.X.X/sample.exe" (native dlls) and set current directory to "app-X.X.X/sample.exe" (I have yet to confirm this would work as expected, but I imagine so 🤞 )

@paulcbetts What are your thoughts on the matter? Am I´m overthinking this? Feedback appreciated 🍻

**Comment by paulcbetts** [comment]
@peters Are you using the latest version of Squirrel.Windows and launching via the execution stubs? I want to confirm the current Best Practice approach doesn't work. If not, Plan 2 might work, or we can try:

3 - Explicitly add `app-X.X.X` to the LoadLibrary path, then call LoadLibrary on ThatExe.exe and call into the import with ordinal zero which is guaranteed to be their version of main(). That might Get Weird though because we're not launching them into a pristine environment, we're launching them into one where msvcrt is already initialized etc etc etc.

**Comment by peters** [comment]
@paulcbetts Yes, I'm using the latest version. Tested on Windows 10 Pro/Enterprise (Anniversary / Creators Update). Firewall occur just after the `CreateProcess` call. I haven't tested any older Windows versions.

>  If not, Plan 2 might work

How about the following (Mix of plan 2 and 3):

1. Copy `%LOCALAPPDATA%\sample\app-X.X.X\sample.exe` to `%LOCALAPPDATA%\sample\sample.exe`
2. Create `sample.exe.config`

```
<configuration>
  <runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <probing privatePath="bin\DLLs" />
    </assemblyBinding>
  </runtime>
</configuration>
```

3. [LoadLibrary approach](https://www.codeproject.com/Articles/1045674/Load-EXE-as-DLL-Mission-Possible)

Let's hope there aren't too many caveats using the LoadLibrary approach. If that's the case I guess we could signal the executable with a new CLI argument called `squirrel-cwd=%LOCALAPPDATA%\app-X.X.X` and require target application to change their current working directory.

**Comment by jonathanbp** [comment]
Any new developments on getting this fixed?

**Comment by paulcbetts** [comment]
Nobody is working on this afaik, if you have any good ideas / insider info on how to convince the Firewall that our app is The Same App, we'd definitely take a PR to fix this bug

**Comment by jonoomph** [comment]
Just curious if there are any updates to this ticket? Any known workarounds for the Firewall prompt on every update? Thanks!

**Comment by rrhornback** [comment]
Any progress or updates to report?

**Comment by JohnL33** [comment]
Looks like https://github.com/Squirrel/Squirrel.Windows/pull/951 is the best bet? Or at least something along those lines.
1. Check for update/initial (initial install just names the folder to "current" or something) - (squirrel is running in current)
2. Download update to app-x.x.x - (squirrel is running in current)
3. Copy app-x.x.x to temp - (squirrel is running in current)
4. Relaunch into app-x.x.x - (squirrel is now running in app-x.x.x)
5. copy temp to current - (squirrel is running in app-x.x.x)
6. Delete temp and relaunch into current

**Comment by sandeep1995** [comment]
We are still stuck on this issue. @shiftkey Can you take a look at this please?

**Comment by sandeep1995** [comment]
Do we have a solution for this yet?

**Comment by SultanaJon** [comment]
Any update on this?

**Comment by Rambarani** [comment]
@peters I also facing this issue. did you able to fix this ?

**Comment by SultanaJon** [comment]
@Rambarani I don't this this is fixed. Microsoft Teams fixed this by creating a `previous` and `current` directory where their previously used version will live and their current (new) version will live. This allows them to add an allowed firewall exception to the exe in the `current` directory and allows their app to work when updated.

**Comment by caesay** [comment]
This has been fixed in my Squirrel fork (tracked in https://github.com/clowd/Clowd.Squirrel/issues/24) by following a similar approach to how msft teams did it with their proprietary fork. This solved loads of issues, tray icon pinning, firewall rules, gpu affinity, taskbar pinned shortcuts, etc. It will be released in v3.0 (eta early June).

**Comment by Rambarani** [comment]
> @Rambarani I don't this this is fixed. Microsoft Teams fixed this by creating a `previous` and `current` directory where their previously used version will live and their current (new) version will live. This allows them to add an allowed firewall exception to the exe in the `current` directory and allows their app to work when updated.

Thanks @SultanaJon

**Comment by Rambarani** [comment]
> This has been fixed in my Squirrel fork (tracked in [clowd#24](https://github.com/clowd/Clowd.Squirrel/issues/24)) by following a similar approach to how msft teams did it with their proprietary fork. This solved loads of issues, tray icon pinning, firewall rules, gpu affinity, taskbar pinned shortcuts, etc. It will be released in v3.0 (eta early June).

Thanks @caesay
